server/src/main/java/org/apache/druid/metadata/SqlSegmentsMetadataQuery.java

```diff
@@ -345,6 +353,42 @@ private CloseableIterator<DataSegment> retrieveSegments(
       @Nullable final Integer limit
   )
   {
+    if (intervals.isEmpty()) {
+      return CloseableIterators.withEmptyBaggage(
+          retrieveSegmentsInIntervalsBatch(dataSource, intervals, matchMode, used, limit).getIterator()
+      );
+    } else {
+      final List<List<Interval>> intervalsLists = Lists.partition(new ArrayList<>(intervals), MAX_INTERVALS_PER_BATCH);
+      final List<Iterator<DataSegment>> resultingIterators = new ArrayList<>();
+      int totalFetched = 0;
+
+      for (final List<Interval> intervalList : intervalsLists) {
+        IteratorWithCount<DataSegment> resultIterator = retrieveSegmentsInIntervalsBatch(dataSource, intervalList, matchMode, used, limit);
+        resultingIterators.add(resultIterator.getIterator());
+        totalFetched += resultIterator.getCount();
+
+        if (null != limit && totalFetched >= limit) {
+          break;
```

Comment:
If a user adds a version of this function that uses multiple intervals with limit, I think that we can return an iterator with more segments than what the limit specifies, breaking the implied contract of the function. Can we cap the returned iterator at the limit? There is logic in `KillUnusedSegmentsTask` that relies heavily on the limit being absolute maximum of segments returned. Maybe we can only add the iterator to the resulteringIterators if it's added size wouldn't cause the limit to be violated?

Reply:
Yes, thanks. The updated code now strictly accounts for limit and added unit tests.